Docs: Update programming FAQ (#144573)
python/cpython: Doc/faq/programming.rst

   .. contents::

General questions
=================

> Is there a source-code level debugger with breakpoints and single-stepping?

Reviewer: ```suggestion
Is there a source code-level debugger with breakpoints and single-stepping?
```

I _think_ the hyphen should be between code and level, since source code is the noun?

Author: Hmm, yeah, we've lots of noun phrases compounding here:

"[[source code] level] debugger"

We could perhaps even do "source-code-level debugger" but I think "source code-level debugger" is better than my original, let's do that.

for pdb as an example.

The IDLE interactive development environment, which is part of the standard

> Python distribution (normally available as :source:`Tools/scripts/idle3`),

Reviewer 2: ```suggestion
Python distribution (normally available as :mod:`idlelib`),
```
[The docs](https://docs.python.org/3/library/idle.html#command-line-usage) seem to prefer invoking the module.

This happens because ``x`` is not local to the lambdas, but is defined in
the outer scope, and it is accessed when the lambda is called --- not when it
is defined.  At the end of the loop, the value of ``x`` is ``4``, so all the

> functions now return ``4**2``, or ``16``.  You can also verify this by

Reviewer 3: Can we use "that is" instead of "or". I got a bit confused as "or" was not followed by "equivalently".

Author: Yes, good idea, that's clearer.

Remember that arguments are passed by assignment in Python.  Since assignment
just creates references to objects, there's no alias between an argument name in

> the caller and callee, and so no call-by-reference as such.  You can achieve the

Reviewer 3: Maybe 

```suggestion
the caller and callee, and hence, no call-by-reference. You can achieve the
```

Author: I'm not keen on "hence", how about "for this reason" or "as a result"?

But is "hence" slightly changing the meaning? "per se" is hedging, saying by itself, there's no call-by-reference.

"hence" is saying one thing leads to the other.

Reviewer 3: Yes, and I think it's fine to have the meaning of "hence" because that's how I understood the "and so [...] as such". Maybe, "In particular, there is no call-by-reference"?

Reviewer 2: >  "as a result"?

I agree "hence" is fine (but the comma is incorrect IMO), but this suggestion also works.

Author: ```suggestion
the caller and callee, and so no call-by-reference as a result. You can achieve the
```

Reviewer 3: I would have used "and consequently no call-by-reference" instead of the "and so [...] as a result", but I assume that you're also not very fond of "consequently".

Author: Sure, that's fine!
```suggestion
the caller and callee, and consequently no call-by-reference.  You can achieve the
```

This answer actually applies to all methods, but the question usually comes up
first in the context of constructors.

In C++ you'd write:

> .. code-block:: c++

Reviewer 3: Do we use `c++` or `cpp` for such code-blocks in general?

Reviewer 2: Neither, we don't have any!

```
$ grep -r ".. code-block:: c++"
$ grep -r ".. code-block:: cpp"
```
